### PR TITLE
Copy mutable variables on assignment (configuration_data and environment)

### LINKED ISSUE #831: configuration_data: assignment to new variable does not make a copy
The way this seems to work currently seems counterintuitive to me, I don't know if it's intentional or a bug, so I filed an issue.

In short, I think this should work:

```
a = configuration_data()
a.set('HELLO', 1)

b = a
configure_file(output : 'a.h', configuration : b)

# This should still work, as we didn't use the original above but a copy!
a.set('WORLD', 1)
```

But it results in a `Can not set values on configuration object that has been used.` error.

I'd attach a patch for the unit tests, but github won't let me ("We don't support that file type.")!

### PR DESCRIPTION
**Pull request by tp-m**
All assignments in meson should be by value, so mutable objects
(i.e. environment() and configuration_data()) should be copied
automatically on assignment.
    
Fixes #831

### CODE REVIEW

**Comment by nirbheek** [comment]
Can we also get a test that does the same for the environment object? :)

File: mesonbuild/interpreter.py [hunk 1]
```diff
@@ -2460,6 +2456,9 @@ def assignment(self, node):
         value = self.to_native(value)
         if not self.is_assignable(value):
             raise InvalidCode('Tried to assign an invalid value to variable.')
+        # For mutable objects we need to make a copy on assignment
+        if isinstance(value, EnvironmentVariablesHolder) or isinstance(value, ConfigurationDataHolder):
+            value = copy.deepcopy(value)
```
Comment (nirbheek) [inline mesonbuild/interpreter.py]: I'm wondering if we should add a `mutable` flag on various variable holders and enforce that everywhere.

**Comment by tp-m** [comment]
Instead of a flag I've added a MutableInterpreterObject class, which seemed the easiest since we have to check the instance type anyway if I'm not mistaken. (I might have misunderstood what you asked for though.)

I have added a test for the `environment()` object being copied properly as well.

**Comment by nirbheek** [comment]
The new class actually makes more sense. LGTM.

**Pull request merged**